dawn/src/dawn/Optimizer/PassSetCaches.cpp

```diff
@@ -248,6 +249,32 @@ bool PassSetCaches::run(const std::shared_ptr<iir::StencilInstantiation>& instan
                field.getExtents().isPointwise())))
             continue;
 
+          // If the Field about to be cached is in a stage with a global index
+          // it can in fact only be chached if all subsequent stages have compatible iteration
+          // spaces
+
+          //find first stage in which the field is used
+          int stageIter = 0;
+          for (const auto& stage : ms.getChildren()) {
+            if (stage->getFields().count(field.getAccessID()) && stage->hasIterationSpace()) {
+              break;
+            }
+            stageIter++;
+          }
+
+          //check if subsequent uses are compatible
+          auto firstUsageWithGlobalIndex = ms.getChildren().begin();
+          std::advance(firstUsageWithGlobalIndex, stageIter);
+          bool incompatibleIterationSpaces = false;
+          for (auto it = firstUsageWithGlobalIndex; it != ms.getChildren().end(); it++) {
```

Comment:
is this loop over all other stage? It seems like the check that the field in question is used is missing?

Reply:
loop is over all subsequent stages. that check that field is in use was indeed missing, good catch!